plugins/liquid/refined-types-plugin/src/main/kotlin/arrow/meta/plugins/liquid/phases/analysis/solver/collect/ConstraintCollection.kt

```diff
@@ -185,30 +192,135 @@ internal fun CompilerContext.collectDeclarationsConstraints(
  */
 internal fun KtDeclaration.constraints(
   solverState: SolverState,
-  context: DeclarationCheckerContext
-): List<Pair<ResolvedCall<*>, NamedConstraint>> =
-  constraintsDSLElements().filterIsInstance<KtElement>().mapNotNull { element ->
-    val bindingCtx = context.trace.bindingContext
-    val call = element.getResolvedCall(bindingCtx)
-    if (call?.preOrPostCall() == true) {
-      val predicateArg = call.arg("predicate")
-      val result = solverState.solver.expressionToFormula(predicateArg, bindingCtx) as? BooleanFormula
-      if (result == null) {
-        context.trace.report(
-          MetaErrors.ErrorParsingPredicate.on(element, ErrorMessages.Parsing.errorParsingPredicate(predicateArg))
-        )
-        solverState.signalParseErrors()
-        null
-      } else {
-        val msgBody = call.arg("msg")
-        val msg = if (msgBody is KtLambdaExpression) msgBody.bodyExpression?.firstStatement?.text?.trim('"')
-        else msgBody?.text ?: predicateArg?.text
-        msg?.let { call to NamedConstraint(it, result) }
+  context: DeclarationCheckerContext,
+  descriptor: DeclarationDescriptor
+) {
+  when (this) {
+    is KtConstructor<*> -> {
+      val constraints = constraintsFromConstructor(solverState, context)
+      Pair(arrayListOf<NamedConstraint>(), arrayListOf<NamedConstraint>())
+      val preConstraints = arrayListOf<NamedConstraint>()
+      val postConstraints = arrayListOf<NamedConstraint>()
+      // rewrite formulae from constructors
+      constraints.forEach { (call, formula) ->
+        if (call.preCall()) {
+          val rewritten = rewritePrecondition(solverState, context, call, formula.formula)
+          preConstraints.add(NamedConstraint(formula.msg, rewritten))
+        }
+        // in constructors 'require' has a double duty
+        if (call.postCall() || call.requireCall()) {
+          val rewritten = rewritePostcondition(solverState, formula.formula)
+          postConstraints.add(NamedConstraint(formula.msg, rewritten))
+        }
       }
```

Comment:
We should probably move these implementations inside the when branches to separate functions that hint at what is done on each case, but we can do that if you want in another cleanup PR

Reply:
Done!